### PR TITLE
Changes to allow Windows backspace keybindings

### PR DESCRIPTION
**Pull request by fdncred**
Added the ability to use `Ctrl+Backspace`, `Meta+Backspace`, and `Shift+Backspace` as keybindings on Windows. Thanks for your help @gwenn. I figure you'll have changes and not like some of the things I've done. :) I didn't run `cargo fmt --all` because I'm not sure if that's your practice.

### CODE REVIEW

File: examples/example.rs [hunk 1]
```diff
@@ -83,7 +83,7 @@ fn main() -> rustyline::Result<()> {
     let config = Config::builder()
         .history_ignore_space(true)
         .completion_type(CompletionType::List)
-        .edit_mode(EditMode::Emacs)
+        .edit_mode(EditMode::Vi)
```
Comment (gwenn) [inline examples/example.rs:86]: Please keep default `Emacs` mode.

File: examples/example.rs [hunk 2]
```diff
@@ -97,6 +97,9 @@ fn main() -> rustyline::Result<()> {
     rl.set_helper(Some(h));
     rl.bind_sequence(KeyPress::Meta('N'), Cmd::HistorySearchForward);
     rl.bind_sequence(KeyPress::Meta('P'), Cmd::HistorySearchBackward);
+    rl.bind_sequence(KeyPress::ControlBackspace, Cmd::Kill(Movement::BackwardWord(1, Word::Vi)));
+    rl.bind_sequence(KeyPress::MetaBackspace, Cmd::Kill(Movement::BackwardChar(1)));
+    rl.bind_sequence(KeyPress::ShiftBackspace, Cmd::Kill(Movement::BackwardChar(1)));
```
Comment (gwenn) [inline examples/example.rs:102]: As these bindings don't work on unix, I would prefer that you remove them.

File: src/tty/windows.rs [hunk 3]
```diff
@@ -179,6 +179,17 @@ impl RawReader for ConsoleRawReader {
...
+                        return Ok(if ctrl {
+                            KeyPress::ControlBackspace
+                        } else if shift {
+                            KeyPress::ShiftBackspace
+                        } else if meta {
+                            KeyPress::MetaBackspace
+                        } else {
+                            KeyPress::Backspace
+                        });
+                    }
```
Comment (gwenn) [inline src/tty/windows.rs:192]: These lines are partially redundant with lines 235:236 and 247:248. So I would prefer you (re)move this block.

File: src/tty/windows.rs [hunk 4]
```diff
@@ -220,13 +231,21 @@ impl RawReader for ConsoleRawReader {
                 };
                 let c = rc?;
                 if meta {
-                    return Ok(KeyPress::Meta(c));
+                    let mut key = keys::char_to_key_press(c);
```
Comment (gwenn) [inline src/tty/windows.rs:234]: You can factorize duplicated lines 234 and 242 by moving it at line 233.

File: src/keys.rs [hunk 5]
```diff
@@ -24,6 +24,8 @@ pub enum KeyPress {
     ControlRight,
     /// Ctrl-↑
     ControlUp,
+    /// Control Backspace
```
Comment (gwenn) [inline src/keys.rs:27]: Maybe we should warn that this key can/may not be used on unix.

File: examples/example.rs (hunk 1 above)
Comment (fdncred) [reply examples/example.rs:86]: Understood.

File: examples/example.rs (hunk 2 above)
Comment (fdncred) [reply examples/example.rs:102]: I left those in there so they could be tested with example.rs, but I'll remove them

File: src/tty/windows.rs (hunk 3 above)
Comment (fdncred) [reply src/tty/windows.rs:192]: understood.

File: src/tty/windows.rs (hunk 4 above)
Comment (fdncred) [reply src/tty/windows.rs:234]: understood.

File: src/keys.rs (hunk 5 above)
Comment (fdncred) [reply src/keys.rs:27]: I wasn't sure if you wanted that comment on ControlBackspace only or ShiftBackspace and MetaBackspace, so I only put it where you called it out.

**Comment by fdncred** [comment]
I believe I changed all your requested items. Please let me know if you have further feedback.

File: src/tty/windows.rs [hunk 6]
```diff
@@ -219,14 +219,21 @@ impl RawReader for ConsoleRawReader {
...
+                    }
+                    return Ok(key);
                 } else {
-                    let mut key = keys::char_to_key_press(c);
                     if key == KeyPress::Tab && shift {
                         key = KeyPress::BackTab;
                     } else if key == KeyPress::Char(' ') && ctrl {
                         key = KeyPress::Ctrl(' ');
+                    } else if key == KeyPress::Backspace && ctrl {
+                        key = KeyPress::ControlBackspace;
```
Comment (gwenn) [inline src/tty/windows.rs:236]: You forgot `ShiftBackspace` case.

File: src/lib.rs [hunk 7]
```diff
@@ -888,7 +888,22 @@ impl<H: Helper> Editor<H> {
...
-            bindings.insert(key_seq, cmd)
+            let key = if let KeyPress::Char(ref c) = key_seq {
+                if c.is_control() {
+                    keys::char_to_key_press(*c)
+                } else {
+                    key_seq
+                }
+            } else if let KeyPress::Ctrl(ref c) = key_seq {
+                if c.is_control() {
+                    keys::char_to_key_press(*c);
```
Comment (gwenn) [inline src/lib.rs:899]: This line can be removed.

File: src/tty/windows.rs (hunk 6 above)
Comment (fdncred) [reply src/tty/windows.rs:236]: understood.

File: src/lib.rs (hunk 7 above)
Comment (fdncred) [reply src/lib.rs]: understood.

File: src/tty/windows.rs [hunk 8]
```diff
@@ -219,14 +219,23 @@ impl RawReader for ConsoleRawReader {
                     return Err(error::ReadlineError::Eof);
                 };
                 let c = rc?;
+                let mut key = keys::char_to_key_press(c);
                 if meta {
-                    return Ok(KeyPress::Meta(c));
+                    if key == KeyPress::Backspace && meta {
```
Comment (gwenn) [inline src/tty/windows.rs:224]: `meta` is always `true`..

File: src/lib.rs [hunk 9]
```diff
@@ -888,7 +888,21 @@ impl<H: Helper> Editor<H> {
...
-            bindings.insert(key_seq, cmd)
+            let key = if let KeyPress::Char(ref c) = key_seq {
+                if c.is_control() {
+                    keys::char_to_key_press(*c)
+                } else {
+                    key_seq
+                }
+            } else if let KeyPress::Ctrl(ref c) = key_seq {
+                if c.is_control() {
+                    warn!(target: "rustyline", "KeyPress::Ctrl({:?}) may not work on unix", c)
```
Comment (gwenn) [inline src/lib.rs:899]: In fact, based on current implementation, both unix and windows do not support such combination...

File: src/tty/windows.rs (hunk 8 above)
Comment (fdncred) [reply src/tty/windows.rs]: done

File: src/lib.rs (hunk 9 above)
Comment (fdncred) [reply src/lib.rs:899]: I added these three items back in the example program and tested it and it performed as expected.
```rust
    rl.bind_sequence(KeyPress::ControlBackspace, Cmd::Kill(Movement::BackwardWord(1, Word::Big)));
    rl.bind_sequence(KeyPress::MetaBackspace, Cmd::Kill(Movement::BackwardWord(1, Word::Big)));
    rl.bind_sequence(KeyPress::ShiftBackspace, Cmd::Kill(Movement::BackwardWord(1, Word::Big)));
```
Please help me understand what you're talking about when you say, `both unix and windows do not support such a combination`

Comment (gwenn) [reply src/lib.rs:899]: Sorry, I mean `KeyPress::Crtl('\x08')`, for example, does not work on unix or windows...

Comment (fdncred) [reply src/lib.rs:899]: It's not supposed to work in this PR. That's why `ControlBackspace` was added.

Comment (gwenn) [reply src/lib.rs:899]: So the comment is misleading.

Comment (fdncred) [reply src/lib.rs:899]: I'm sorry @gwenn but I'm not following what you're saying. What comment is misleading? I was originally trying to add `KeyPress:Ctrl('\x08')` but you made it clear in this PR that it wouldn't work that way. So, I added `ControlBackspace`, `MetaBackspace`, `ShiftBackspace` in place of being able to take char literals.

Comment (gwenn) [reply src/lib.rs:899]: At line 899:
`KeyPress::Ctrl({:?}) may not work on unix`
should be rephrased:
`KeyPress::Ctrl({:?}) does not work`
or something like this.

Comment (fdncred) [reply src/lib.rs]: updated

**Comment by fdncred** [comment]
@gwenn ping. are we ready to merge this?

**Comment by gwenn** [comment]
I would prefer to fix and merge #333 before.

**Pull request closed without merging**